[Card Game 3] Задачи 4, 5, долг 1

SenX51 opened the pull request:
**Выполнил:**
 > Семенов А.В.

**Задача 4:**
 > Завершить игровую логику. Проект должен быть играбельным.
Включает в себя. Реализацию необходимого минимума игровых возможностей:
Полный цикл жизни карты (Колода -> Добор -> Рука -> Сброс -> Добор).
Направленные на противника карты должны требовать цель, ненаправленные иметь возможность быть разыгранными без цели.
Возможность нанесения урона персонажу противника с помощью карт.
Возможность нанесения урона персонажу игрока противником.
Возможность наложения защиты на персонажа игрока.
Возможность персонажу противника накладывать защиту на себя.
Реализовать пошаговую систему боя. В свой ход игрок может разыгрывать карты, при завершении противник выполняет действия по алгоритму.
Отслеживание состояния игры (противники побеждены, игрок повержен).

QOL: 
Переписана система разрешения игровых эффектов (разыгрывание карт, применение индивидуальных действий карт, завершение раунда и пр.) из спагетти асинхронных методов/промисов на систему с буффером, в который кладется последовательность действий. Если она может выполниться - выполняется, если что-то идет не так - выполнение прерывается.
__
**Задача 5:**
 > Привести игру к **полностью** играбельному формату - имплементировать все игровые возможности. Дать пользователю инструменты, чтобы использовать все разработанные фичи. Геймплейный цикл (gameplay loop) должен быть полноценным.

Игровой цикл:
Меню -> Начало игры -> Выбор сражения из нескольких доступных -> Сражение -> Получение и выбор наград -> Выбор следующего боя.
В случае смерти игрока - завершение текущего сражения.

Удалось переиспользовать компонент CardPile (просмотр стопок карт) для системы выбора наград.
 __
**Долг по задаче 1:**
 > Доделал анимации взятия карт в руку через TransitionGroup. Суммарно на анимирование карт потрачено минимум часов 8 работы. Оказывается, если в группе активна ЛЮБАЯ анимация - все элементы TransitionGroup теряют вычисляемые стили (:style).
__
**Комментарии**
 > 21.12.2024 4:25: Обновил PR. Добавил в этот PR 5 лабораторную и долги, чтобы изменения из этого PR не подтягивались в новый. Этот был формально зачтен, все правки были с пометкой "подумать, можно ли улучшить", поэтому посчитал целесообразным добавить все обновления сюда.
___
**Демонстрация:**
### [**Видео. Задача 4**](https://github.com/user-attachments/assets/3a8e60f7-bbcd-4667-be2a-5d3860d44025)
### [**Видео. Задача 5 + Долг**](https://github.com/user-attachments/assets/43a46903-23e3-456d-8e41-d77f5d22fa7e)

![image](https://github.com/user-attachments/assets/dde4572d-0f86-4c54-baf5-573d4813affe)

SenX51 committed:
Merge pull request #1 from SenX51/card-game-3-enemy

Merge Enemies

SenX51 committed:
Combat System

> Encounter emits event after every enemy has died and repeats itself
> Enemies use weighted random to choose pattern
> Enemies execute chosen pattern step by step and choose new one when it ends
> Cards execute their actions
> Updated UI

reznikovkg commented on vue/src/components/enemy/EnemyComponent.vue:
```diff
@@ -4,14 +4,14 @@
-        x{{ enemy.currentPattern[0].values[1] }}
+      <p>{{ enemy.intent.values[0] }}</p>
+      <p v-if="enemy.intent.type == actions.ATTACK_MULTIPLE">
+        x{{ enemy.intent.values[1] }}
```
Лучше переделать на объект с ключами

reznikovkg commented on vue/src/components/pages/BattlePage.vue:
```diff
@@ -73,36 +93,184 @@ export default {
+    }
+  },
+  watch: {
+    enemies: function (newVal) {
```
Лучше не использовать watch

SenX51 committed:
Logic Rewrite

 > Rewritten Async/Await/Promise mixed intertwined spaghetti to a structured buffer
 > Added attack and defend SFX
 > Waisted yet another 8 hours of life trying to once again implement animations. Turns out if ANY animation is active inside of transition-group - all of dynamic :style simply doesn't work. I'm too done with this feature to rewrite all of the code to make up a workaround 😫
 > Updated some CSS

SenX51 committed:
Menu Page

SenX51 committed:
Fully Finished Build

Added:
 > Encounter choice
 > Reward choice
 > Game automatically reboots on death
Update:
 > int rarity -> Enum
 > Enemies/Cards filter default values on Get

reznikovkg commented:
Сдано
